### PR TITLE
add subscription deletion workflow

### PR DESCRIPTION
**Pull request by chamilaadhi**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java [hunk 1]
```diff
@@ -145,11 +146,35 @@ private Workflow createWorkflowFromResultSet(ResultSet rs) throws SQLException,
...
+        final String getworkflowQuery = "SELECT * FROM AM_WORKFLOWS WHERE WF_REFERENCE=? AND WF_TYPE=?";
+  
+        Workflow workflow;
+        try (Connection conn = DAOUtil.getConnection();
+                PreparedStatement ps = conn.prepareStatement(getworkflowQuery)) {
+            ps.setString(1, subscriptionId);
+            ps.setString(2, APIMgtConstants.WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION);
+            try (ResultSet rs = ps.executeQuery()) {
+                workflow = this.createWorkflowFromResultSet(rs);
+
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: Remove this blank line.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.wso2.carbon.apimgt.core.exception.WorkflowException;
+import org.wso2.carbon.apimgt.core.models.Workflow;
+import org.wso2.carbon.apimgt.core.models.WorkflowStatus;
+
+/**
+ * Subscription deletion simple workflow
+ */
+public class SubscriptionDeletionSimpleWorkflowExecutor implements WorkflowExecutor {
+    private static final Log log = LogFactory.getLog(SubscriptionDeletionSimpleWorkflowExecutor.class);
+
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java]: Add comments to the public methods

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java (hunk 1 above)
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/WorkflowDAOImpl.java]: done

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java (hunk 2 above)
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java]: done

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 3]
```diff
@@ -633,50 +677,65 @@ public void completeWorkflow(WorkflowExecutor workflowExecutor, Workflow workflo
...
+                    log.debug("Subscription Creation workflow complete: Rejected");
                 }
                 subscriptionState = APIMgtConstants.SubscriptionStatus.REJECTED;
             }
 
             getApiSubscriptionDAO().updateSubscriptionStatus(workflow.getWorkflowReference(), subscriptionState);
+        } else if (workflow instanceof SubscriptionWorkflow
+                && WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_DELETION.equals(workflow.getWorkflowType())) {
+            WorkflowResponse response = workflowExecutor.complete(workflow);
+            if (WorkflowStatus.APPROVED == response.getWorkflowStatus()) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: WorkflowStatus is an enum and can use '==' to do comparison

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 4]
```diff
@@ -264,58 +260,107 @@ public SubscriptionResponse addApiSubscription(String apiId, String applicationI
...
+            Subscription subscription = getApiSubscriptionDAO().getAPISubscription(subscriptionId);
+            if (subscription == null) {
+                String errorMsg = "Subscription not found for the id - " + subscriptionId;
+                log.error(errorMsg);
+                throw new APIManagementException(errorMsg, ExceptionCodes.SUBSCRIPTION_NOT_FOUND);
+            } else {
+                if (APIMgtConstants.SubscriptionStatus.ON_HOLD == subscription.getStatus()) {
+                    String pendingRefForSubscription = getWorkflowDAO()
+                            .getExternalWorkflowReferenceForSubscription(subscriptionId);
+                    if (pendingRefForSubscription != null) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: if you take the overall feature, this reference number is generated by the code. It is a random number. so this value is either null (if there is no entry in db) or some random value. There is no chance of this value becoming an empty one.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * execute subscription deletion 
+     */
+    @Override
+    public WorkflowResponse execute(Workflow workflow) throws WorkflowException {
+        if (log.isDebugEnabled()) {
+            log.info("Executing Subscription deletion Workflow..");
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java]: removed

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        return workflowResponse;
+    }
+
+    /**
+     * complete subscription deletion 
+     */
+    @Override
+    public WorkflowResponse complete(Workflow workflow) throws WorkflowException {
+        if (log.isDebugEnabled()) {
+            log.info("Complete  Subscription deletion Workflow..");
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/workflow/SubscriptionDeletionSimpleWorkflowExecutor.java]: removed

File: components/apimgt/org.wso2.carbon.apimgt.core/src/test/java/org/wso2/carbon/apimgt/core/impl/APIStoreImplTestCase.java [hunk 7]
```diff
@@ -306,16 +310,24 @@ public void testAddSubscriptionForInvalidAPI() throws APIManagementException {
...
         APISubscriptionDAO apiSubscriptionDAO = mock(APISubscriptionDAO.class);
+        WorkflowDAO workflowDAO = mock(WorkflowDAO.class);
         APIStore apiStore = new APIStoreImpl(USER_NAME, null, applicationDAO, apiSubscriptionDAO, null, null, null,
-                null);
+                workflowDAO);   
+        
+        Application application = SampleTestObjectCreator.createDefaultApplication();
+        APIBuilder builder = SampleTestObjectCreator.createDefaultAPI();
+        API api = builder.build();
+       // API api = mock(API.class);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/test/java/org/wso2/carbon/apimgt/core/impl/APIStoreImplTestCase.java]: removed

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java (hunk 4 above)
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: okey :+1:

**Pull request merged**